lib/analyzer/applications/java.ts

```diff
@@ -218,29 +215,41 @@ function unpackJars(
   const requiredLevelsOfUnpacking = desiredLevelsOfUnpacking + 1;
   const fingerprints: JarFingerprint[] = [];
 
+  // jarBuffers is the array of JARS found in the image layers;
+  // this represents the 1st "level" which we will unpack by
+  // default to analyse. Any JARs found when analysing are nested
+  // and we will keep going until we have no more nested JARs or
+  // the desired level of unpacking is met
   for (const jarBuffer of jarBuffers) {
-    const unpackedLevels: number = 0;
-    const jars: JarBuffer[] = unpackJarsTraverse({
-      jarBuffer: jarBuffer.digest,
+    const jarInfo = unpackJar({
+      jarBuffer: jarBuffer.buffer,
       jarPath: jarBuffer.location,
       desiredLevelsOfUnpacking,
       requiredLevelsOfUnpacking,
       unpackedLevels,
-      coords: null,
-      jarBuffers: [],
     });
 
     // if any of the coords are null for this JAR we didn't manage to get
     // anything from the JAR's pom.properties manifest, so calculate the
     // sha so maven-deps can fallback to searching maven central
-    jars.forEach((jar) => {
-      fingerprints.push({
-        location: jar.location,
-        digest: jar.coords ? null : bufferToSha1(jar.digest),
-        dependencies: jar.dependencies,
-        ...jar.coords,
-      });
+    fingerprints.push({
+      location: jarInfo.location,
+      digest: jarInfo.coords ? null : bufferToSha1(jarInfo.buffer),
+      dependencies: jarInfo.dependencies,
+      ...jarInfo.coords,
     });
+
+    if (jarInfo.nestedJars.length > 0) {
+      // this is an uber/fat JAR so we need to unpack the nested JARs to
+      // analyse them for coords and further nested JARs (depth flag allowing)
+      fingerprints.push(
+        ...unpackJars(
+          jarInfo.nestedJars as JarBuffer[],
```

Comment:
does it really have all the properties of JarBuffer, or is it a partial? would it make more sense to change line 202 to `Partial<JarBuffers>[]` (if that's a thing...)? not sure I fully follow and understand the source of the type difference...

Reply:
This was super unclear too. I've reduced the scope of the `JarBuffer` type and extended the `JarInfo` type. Now `unpackJars` method takes a `JarBuffer[]` and `unpackJar` method returns `JarInfo` which has the decorated information needed to create a `Fingerprint` type.